pkg/activitypub/httpsig/algorithm.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package httpsig
+
+import (
+	"crypto/ed25519"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/hyperledger/aries-framework-go/pkg/crypto"
+	ariesverifier "github.com/hyperledger/aries-framework-go/pkg/doc/signature/verifier"
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	httpsig "github.com/igor-pavlenko/httpsignatures-go"
+)
+
+const orbHTTPSigAlgorithm = "https://github.com/trustbloc/orb/httpsig"
+
+// ErrInvalidSignature indicates that the signature is not valid for the given data.
+var ErrInvalidSignature = errors.New("invalid HTTP signature")
+
+type keyResolver interface {
+	// Resolve returns the public key bytes and the type of public key for the given key ID.
+	Resolve(keyID string) (*ariesverifier.PublicKey, error)
+}
+
+// SignatureHashAlgorithm is a custom httpsignatures.SignatureHashAlgorithm that uses KMS to sign HTTP requests.
+type SignatureHashAlgorithm struct {
+	Crypto      crypto.Crypto
+	KMS         kms.KeyManager
+	keyResolver keyResolver
+	keyID       string
+}
+
+// NewSignerAlgorithm returns a new SignatureHashAlgorithm which uses KMS to sign HTTP requests.
+func NewSignerAlgorithm(c crypto.Crypto, km kms.KeyManager, keyID string) *SignatureHashAlgorithm {
+	return &SignatureHashAlgorithm{
+		Crypto: c,
+		KMS:    km,
+		keyID:  keyID,
+	}
+}
+
+// NewVerifierAlgorithm returns a new SignatureHashAlgorithm which is used to verify the signature
+// in the HTTP request header.
+func NewVerifierAlgorithm(c crypto.Crypto, km kms.KeyManager, keyResolver keyResolver) *SignatureHashAlgorithm {
+	return &SignatureHashAlgorithm{
+		Crypto:      c,
+		KMS:         km,
+		keyResolver: keyResolver,
+	}
+}
+
+// Algorithm returns this algorithm's name.
+func (a *SignatureHashAlgorithm) Algorithm() string {
+	return orbHTTPSigAlgorithm
+}
+
+// Create signs data with the secret.
+func (a *SignatureHashAlgorithm) Create(secret httpsig.Secret, data []byte) ([]byte, error) {
+	kh, err := a.KMS.Get(a.keyID)
+	if err != nil {
+		return nil, fmt.Errorf("get key handle: %w", err)
+	}
+
+	logger.Debugf("Got key handle for key ID [%s]: %+v. Signing ...", secret.KeyID, kh)
+
+	sig, err := a.Crypto.Sign(data, kh)
+	if err != nil {
+		return nil, fmt.Errorf("sign data: %w", err)
+	}
+
+	logger.Debugf("... successfully signed data with keyID from KMS [%s]", a.keyID)
+
+	return sig, nil
+}
+
+// Verify verifies the signature over data with the secret.
+func (a *SignatureHashAlgorithm) Verify(secret httpsig.Secret, data, signature []byte) error {
+	pubKey, err := a.keyResolver.Resolve(secret.KeyID)
+	if err != nil {
+		return fmt.Errorf("resolve key %s: %w", secret.KeyID, err)
+	}
+
+	logger.Debugf("Got key %+v from keyID [%s]", pubKey, secret.KeyID)
+
+	if !ed25519.Verify(pubKey.Value, data, signature) {
```

Comment:
Is it always ed25519?

Reply:
I'll open a new issue to support multiple algorithms.